src/metabase/driver/hive_like.clj

```diff
@@ -0,0 +1,160 @@
+      ;; if we need to do an offset we have to do nesting to generate a row number and where on that
+      (let [over-clause (format "row_number() OVER (%s)"
+                                (first (hsql/format (select-keys honeysql-query [:order-by])
+                                                    :allow-dashed-names? true
+                                                    :quoting :mysql)))]
+        (-> (apply h/select (map last (:select honeysql-query)))
+            (h/from (h/merge-select honeysql-query [(hsql/raw over-clause) :__rownum__]))
+            (h/where [:> :__rownum__ offset])
+            (h/limit items))))))
+
+;; lots of copy-paste here. consider making some of those non-private instead.
+(defn- exception->nice-error-message ^String [^java.sql.SQLException e]
+  (or (->> (.getMessage e)     ; error message comes back like 'Column "ZID" not found; SQL statement: ... [error-code]' sometimes
+           (re-find #"^(.*);") ; the user already knows the SQL, and error code is meaningless
+           second)             ; so just return the part of the exception that is relevant
+      (.getMessage e)))
+
+(defn do-with-try-catch {:style/indent 0} [f]
+  (try (f)
+       (catch java.sql.SQLException e
+         (log/error (jdbc/print-sql-exception-chain e))
+         (throw (Exception. (exception->nice-error-message e))))))
+
+(defn- run-query
+  "Run the query itself."
+  [{sql :query, params :params, remark :remark} connection]
+  (let [sql              (str "-- " remark "\n" (hx/unescape-dots sql))
+        statement        (into [sql] params)
+        [columns & rows] (jdbc/query connection statement {:identifiers identity, :as-arrays? true})]
+    {:rows    (or rows [])
+     :columns columns}))
+
+(defn run-query-without-timezone [driver settings connection query]
+  (run-query query connection))
+
+(defprotocol ^:private IUnprepare
+  (^:private unprepare-arg ^String [this]))
+
+(extend-protocol IUnprepare
+  nil     (unprepare-arg [this] "NULL")
+  String  (unprepare-arg [this] (str \' (s/replace this "'" "\\\\'") \')) ; escape single-quotes
+  Boolean (unprepare-arg [this] (if this "TRUE" "FALSE"))
+  Number  (unprepare-arg [this] (str this))
+  Date    (unprepare-arg [this] (first (hsql/format
+                                        (hsql/call :from_unixtime
+                                                   (hsql/call :unix_timestamp
+                                                              (hx/literal (u/date->iso-8601 this))
+                                                              (hx/literal "yyyy-MM-dd\\\\'T\\\\'HH:mm:ss.SSS\\\\'Z\\\\'")))))))
+
+(defn unprepare
+  "Convert a normal SQL `[statement & prepared-statement-args]` vector into a flat, non-prepared statement."
+  ^String [[sql & args]]
+  (loop [sql sql, [arg & more-args, :as args] args]
+    (if-not (seq args)
+      sql
+      (recur (s/replace-first sql #"(?<!\?)\?(?!\?)" (unprepare-arg arg))
+             more-args))))
```

Comment:
Also BTW there's a `metabase.driver.generic-sql.util.unprepare` namespace that has this protocol and functions in it already. So there should be no need to copy-pasta the entire namespace when you can just used the shared code